File: api/src/types/events.ts
// Canonical event types matching Go structs (poller/pkg/types/events.go)
// and JSON schemas (schemas/*.json).
// Keep in sync — CI validates via `just gen-types`.

export type EventType = "PacketSent" | "PacketVerified" | "PacketDelivered";

export interface BaseChainEvent {
  event_type: EventType;
  chain_id: number;
  block_number: number;
  tx_hash: string;
  log_index: number;
  block_timestamp: number;
  src_eid: number;
  sender: string;
  nonce: number;
  ingestion_timestamp: number;
}

export interface PacketSentEvent extends BaseChainEvent {
  event_type: "PacketSent";
  dst_eid: number;
  receiver: string;
  guid: string;
}

export interface PacketVerifiedEvent extends BaseChainEvent {
  event_type: "PacketVerified";
  dvn_address: string;
  receiver?: string;
  payload_hash?: string;
}

export interface PacketDeliveredEvent extends BaseChainEvent {
  event_type: "PacketDelivered";
  receiver?: string;
}

export type ChainEvent = PacketSentEvent | PacketVerifiedEvent | PacketDeliveredEvent;

File: poller/pkg/types/events.go
package types

import (
	"github.com/ethereum/go-ethereum/common"
)

// EventType represents the type of LayerZero event.
type EventType string

const (
	EventPacketSent      EventType = "PacketSent"
	EventPacketVerified  EventType = "PacketVerified"
	EventPacketDelivered EventType = "PacketDelivered"
)

// ChainEvent is the canonical event structure published to Redis Streams.
// This must stay in sync with the JSON schemas in schemas/.
//
//	Go struct ──serialize──▶ Redis Stream ──deserialize──▶ TS interface
type ChainEvent struct {
	EventType          EventType      `json:"event_type"`
	ChainID            int            `json:"chain_id"`
	BlockNumber        uint64         `json:"block_number"`
	TxHash             common.Hash    `json:"tx_hash"`
	LogIndex           uint           `json:"log_index"`
	BlockTimestamp     int64          `json:"block_timestamp"`
	SrcEID             uint32         `json:"src_eid"`
	DstEID             uint32         `json:"dst_eid,omitempty"`
	Sender             common.Address `json:"sender"`
	Receiver           common.Address `json:"receiver,omitempty"`
	Nonce              uint64         `json:"nonce"`
	GUID               common.Hash    `json:"guid,omitempty"`
	DVNAddress         common.Address `json:"dvn_address,omitempty"`
	PayloadHash        common.Hash    `json:"payload_hash,omitempty"`
	IngestionTimestamp int64          `json:"ingestion_timestamp"`
}
